docs/source-fabric/advanced/model_parallel/tp.rst

```diff
@@ -0,0 +1,280 @@
+
+    # 2. Configure devices and set the strategy in Fabric
+    fabric = L.Fabric(accelerator="cuda", devices=2, strategy=strategy)
+    fabric.launch()
+
+The strategy takes the custom parallelization function as input.
+No other changes to your training code are necessary at this point.
+Later in the code, when you call ``fabric.setup(model)``, Fabric will apply the ``parallelize_feedforward`` function to the model automatically.
+
+.. collapse:: Full training example (requires at least 2 GPUs).
+
+    .. code-block:: python
+
+        import torch
+        import torch.nn as nn
+        import torch.nn.functional as F
+
+        from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel
+        from torch.distributed.tensor.parallel import parallelize_module
+
+        import lightning as L
+        from lightning.pytorch.demos.boring_classes import RandomDataset
+        from lightning.fabric.strategies import ModelParallelStrategy
+
+
+        class FeedForward(nn.Module):
+            def __init__(self, dim, hidden_dim):
+                super().__init__()
+                self.w1 = nn.Linear(dim, hidden_dim, bias=False)
+                self.w2 = nn.Linear(hidden_dim, dim, bias=False)
+                self.w3 = nn.Linear(dim, hidden_dim, bias=False)
+
+            def forward(self, x):
+                return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+
+        def parallelize_feedforward(model, device_mesh):
+            # Lightning will set up a device mesh for you
+            tp_mesh = device_mesh["tensor_parallel"]
+            # Use PyTorch's distributed tensor APIs to parallelize the model
+            plan = {
+                "w1": ColwiseParallel(),
+                "w2": RowwiseParallel(),
+                "w3": ColwiseParallel(),
+            }
+            parallelize_module(model, tp_mesh, plan)
+            return model
+
+
+        strategy = ModelParallelStrategy(parallelize_fn=parallelize_feedforward)
+        fabric = L.Fabric(accelerator="cuda", devices=2, strategy=strategy)
+        fabric.launch()
+
+        # Initialize the model
+        model = FeedForward(8192, 8192)
+        model = fabric.setup(model)
+
+        # Define the optimizer
+        optimizer = torch.optim.AdamW(model.parameters(), lr=3e-3, foreach=True)
+        optimizer = fabric.setup_optimizers(optimizer)
+
+        # Define dataset/dataloader
+        dataset = RandomDataset(8192, 64)
+        dataloader = torch.utils.data.DataLoader(dataset, batch_size=8)
+        dataloader = fabric.setup_dataloaders(dataloader)
+
+        # Simplified training loop
+        for i, batch in enumerate(dataloader):
+            output = model(batch)
+            loss = output.sum()
+            fabric.backward(loss)
+            optimizer.step()
+            optimizer.zero_grad()
+            fabric.print(f"Iteration {i} complete")
+
+        fabric.print(f"Peak memory usage: {torch.cuda.max_memory_reserved() / 1e9:.02f} GB")
```

Comment:
Why do we not use `torch.cuda.max_memory_allocated()`? Just a naive question how it's different from `torch.cuda.max_memory_reserved()`. Isn't `torch.cuda.max_memory_reserved()` showing a number that is usually larger if you have more space? E.g., it may allocate more memory on a larger GPU than a smaller GPU even though the code could run on both GPUs?

Reply:
You are right yes, it should be allocated.